Use an ephemeral device ID for the Continuity camera when it is not the system preferred

eric-carlson opened the pull request:
#### 0ee91ab89adb35d40fb7557159dfbb5380326156
<pre>
Use an ephemeral device ID for the Continuity camera when it is not the system preferred
<a href="https://bugs.webkit.org/show_bug.cgi?id=247208">https://bugs.webkit.org/show_bug.cgi?id=247208</a>
rdar://100335624

Reviewed by Jer Noble.

A &quot;Continuity Camera&quot;, an iPhone used as a camera for a Macintosh, is listed as an
AVCaptureDevice whenever it is in close enough to the Macintosh to be used, but it is only
the &quot;system preferred camera&quot; when it is in an appropriate position and orientation to be
used as a camera for the device. This means that a script may remember the device ID when
a user has set up the phone to be used as a camera, and then pass it to getUserMedia later
when the phone is near the Macintosh but not usable as a camera - e.g. in the user&apos;s pocket.

To prevent this problem, use a device ID that is only valid for the lifetime of the
current frame when a Continuity Camera is available but is not the system preferred camera.

* LayoutTests/http/tests/media/media-stream/enumerate-devices-ephemeral-id-expected.txt: Added.
* LayoutTests/http/tests/media/media-stream/enumerate-devices-ephemeral-id.html: Added.
* LayoutTests/http/tests/media/media-stream/resources/enumerate-devices-ephemeral-id-iframe.html: Added.

* Source/WTF/wtf/PlatformHave.h: Define HAVE_CONTINUITY_CAMEARA

* Source/WebCore/Modules/mediastream/CanvasCaptureMediaStreamTrack.cpp:
(WebCore::CanvasCaptureMediaStreamTrack::Source::Source): Update for RealtimeMediaSource constructor
changes.
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::MediaDevices::exposeDevices): Take a MediaDeviceHashSalts instead of a String
because there are now two hash salts.
(WebCore::MediaDevices::enumerateDevices): Ditto.
* Source/WebCore/Modules/mediastream/MediaDevices.h:

* Source/WebCore/Modules/mediastream/UserMediaClient.h: Ditto.
* Source/WebCore/Modules/mediastream/UserMediaController.cpp:
* Source/WebCore/Modules/mediastream/UserMediaController.h:
(WebCore::UserMediaController::enumerateMediaDevices): Ditto.

* Source/WebCore/Modules/mediastream/UserMediaRequest.cpp:
(WebCore::UserMediaRequest::allow): Ditto.
* Source/WebCore/Modules/mediastream/UserMediaRequest.h:

* Source/WebCore/Modules/speech/SpeechRecognitionCaptureSource.cpp:
(WebCore::SpeechRecognitionCaptureSource::createRealtimeMediaSource): Update for
RealtimeMediaSource constructor changes.

* Source/WebCore/Modules/webaudio/MediaStreamAudioSource.cpp:
(WebCore::MediaStreamAudioSource::MediaStreamAudioSource): Ditto.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/platform/mediastream/CaptureDevice.h:
(WebCore::CaptureDevice::CaptureDevice): Add `isEphemeral` parameter.
(WebCore::CaptureDevice::setPersistentId):
(WebCore::CaptureDevice::setLabel):
(WebCore::CaptureDevice::isEphemeral const):
(WebCore::CaptureDevice::setIsEphemeral):
(WebCore::CaptureDevice::encode const):
(WebCore::CaptureDevice::decode):
(WebCore::CaptureDevice::isolatedCopy):

* Source/WebCore/platform/mediastream/RealtimeIncomingAudioSource.cpp:
(WebCore::RealtimeIncomingAudioSource::RealtimeIncomingAudioSource): Update for
RealtimeMediaSource constructor changes.

* Source/WebCore/platform/mediastream/RealtimeIncomingVideoSource.cpp:
(WebCore::RealtimeIncomingVideoSource::RealtimeIncomingVideoSource): Ditto.

* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::toSourceType):
(WebCore::RealtimeMediaSource::RealtimeMediaSource): Take a CaptureDevice instead of a
type, name, and device ID. Take a MediaDeviceHashSalts instead of a String for the salts.
(WebCore::RealtimeMediaSource::setPersistentId):
(WebCore::RealtimeMediaSource::initializePersistentId): Calculate persistent and ephemeral
hashed IDs.
(WebCore::RealtimeMediaSource::fitnessDistance): m_hashedID -&gt; hashedID()
(WebCore::RealtimeMediaSource::hashedId const): Consider ephemeral state.
(WebCore::RealtimeMediaSource::deviceIDHashSalts const):
(WebCore::RealtimeMediaSource::setLogger):
(WebCore::RealtimeMediaSource::deviceIDHashSalt const): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:

* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::createMediaStream): Take MediaDeviceHashSalts with the
hash salts.
(WebCore::RealtimeMediaSourceCenter::getDisplayMediaDevices): Ditto.
(WebCore::RealtimeMediaSourceCenter::getUserMediaDevices): Ditto.
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraints): Ditto.
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraintsAfterEnumeration): Ditto.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.h:

* Source/WebCore/platform/mediastream/RealtimeMediaSourceFactory.h: Ditto.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h: Ditto.

* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::RealtimeVideoCaptureSource): Update for RealtimeMediaSource
constructor changes.
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:

* Source/WebCore/platform/mediastream/RealtimeVideoSource.cpp:
(WebCore::RealtimeVideoSource::RealtimeVideoSource): Ditto.

* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::create): Ditto.
(WebCore::DisplayCaptureSourceCocoa::DisplayCaptureSourceCocoa): Ditto.
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.h:

* Source/WebCore/platform/mediastream/ios/CoreAudioCaptureSourceIOS.h:
* Source/WebCore/platform/mediastream/ios/CoreAudioCaptureSourceIOS.mm:
(WebCore::CoreAudioCaptureSourceFactoryIOS::createAudioCaptureSource): Ditto.

* Source/WebCore/platform/mediastream/mac/AVCaptureDeviceManager.mm:
(WebCore::AVCaptureDeviceManager::updateCachedAVCaptureDevices):
(WebCore::toCaptureDevice): Make Continuity Camera ephemeral when it is not the system
preferred camera.
(WebCore::AVCaptureDeviceManager::retrieveCaptureDevices):

* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::create): Update for RealtimeMediaSource constructor changes.
(WebCore::AVVideoCaptureSource::AVVideoCaptureSource): Ditto.

* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSource::create): Ditto.
(WebCore::CoreAudioCaptureSource::createForTesting): Ditto.
(WebCore::CoreAudioCaptureSource::CoreAudioCaptureSource): Ditto.
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
(WebCore::CoreAudioCaptureSourceFactory::createAudioCaptureSource):

* Source/WebCore/platform/mediastream/mac/MockAudioSharedUnit.mm:
(WebCore::MockRealtimeAudioSource::create): Ditto.

* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.h:
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
(WebCore::MockRealtimeVideoSource::create): Ditto.
(WebCore::MockRealtimeVideoSourceMac::createForMockDisplayCapturer): Ditto.
(WebCore::MockRealtimeVideoSourceMac::MockRealtimeVideoSourceMac): Ditto.

* Source/WebCore/platform/mediastream/mac/RealtimeMediaSourceCenterMac.cpp:
* Source/WebCore/platform/mock/MockMediaDevice.h:
(WebCore::MockMediaDevice::captureDevice const): Allow to be ephemeral .
(WebCore::MockMediaDevice::encode const):
(WebCore::MockMediaDevice::decodeMockMediaDevice):
(WebCore::MockMediaDevice::decode):

* Source/WebCore/platform/mock/MockRealtimeAudioSource.cpp:
(WebCore::MockRealtimeAudioSource::create): Update for RealtimeMediaSource constructor changes.
(WebCore::MockRealtimeAudioSource::MockRealtimeAudioSource): Ditto.
* Source/WebCore/platform/mock/MockRealtimeAudioSource.h:

* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices): No devices are ephemeral by default.
(WebCore::MockDisplayCapturer::MockDisplayCapturer): Update for RealtimeMediaSource constructor
changes.
(WebCore::MockRealtimeMediaSourceCenter::setDeviceIsEphemeral):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.h:

* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::create): Ditto.
(WebCore::MockRealtimeVideoSource::MockRealtimeVideoSource): Ditto.
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::setMockMediaDeviceIsEphemeral): Added for testing.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/UIProcess/API/C/WKMockMediaDevice.cpp:
(WKAddMockMediaDevice):
(WKSetMockMediaDeviceIsEphemeral):
* Source/WebKit/UIProcess/API/C/WKMockMediaDevice.h:

* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::createMicrophoneSource):
(WebKit::UserMediaCaptureManagerProxy::createCameraSource):
(WebKit::UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::setMockMediaDeviceIsEphemeral):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/SpeechRecognitionRemoteRealtimeMediaSource.cpp:
(WebKit::SpeechRecognitionRemoteRealtimeMediaSource::SpeechRecognitionRemoteRealtimeMediaSource):
(): Deleted.
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest): Take MediaDeviceHashSalts
for hash salts.
(WebKit::UserMediaPermissionRequestManagerProxy::didCommitLoadForFrame): Clear ephemeral
salt for a frame.
(WebKit::UserMediaPermissionRequestManagerProxy::resetAccess): Ditto.
(WebKit::UserMediaPermissionRequestManagerProxy::ephemeralDeviceHashSaltForFrame): Calculate
or lookup an ephemeral salt for a frame.
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionRequest): Use
MediaDeviceHashSalts.
(WebKit::UserMediaPermissionRequestManagerProxy::platformValidateUserMediaRequestConstraints): Ditto.
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionValidRequest): Ditto.
(WebKit::UserMediaPermissionRequestManagerProxy::enumerateMediaDevicesForFrame): Ditto.
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:

* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
(WebKit::UserMediaPermissionRequestProxy::setDeviceIdentifierHashSalts): *HashSalt -&gt; *HashSalts
(WebKit::UserMediaPermissionRequestProxy::deviceIdentifierHashSalts const): Ditto
(WebKit::UserMediaPermissionRequestProxy::setDeviceIdentifierHashSalt): Deleted.
(WebKit::UserMediaPermissionRequestProxy::deviceIdentifierHashSalt const): Deleted.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame): Let UserMediaPermissionRequestManager know the
frame has reloaded so it can clear the ephemeral salt ID.
(WebKit::WebPageProxy::enumerateMediaDevicesForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::setMockMediaDeviceIsEphemeral):
* Source/WebKit/UIProcess/WebProcessPool.h:

* Source/WebKit/WebProcess/MediaStream/UserMediaPermissionRequestManager.cpp:
(WebKit::UserMediaPermissionRequestManager::userMediaAccessWasGranted):
(WebKit::UserMediaPermissionRequestManager::enumerateMediaDevices):
* Source/WebKit/WebProcess/MediaStream/UserMediaPermissionRequestManager.h:

* Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.cpp:
(WebKit::WebUserMediaClient::enumerateMediaDevices):
* Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::userMediaAccessWasGranted):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setMockMediaDeviceIsEphemeral):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeAudioSource.cpp:
(WebKit::RemoteRealtimeAudioSource::create): Update for RealtimeMediaSource constructor change.
(WebKit::RemoteRealtimeAudioSource::RemoteRealtimeAudioSource): Ditto.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeAudioSource.h:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::RemoteRealtimeMediaSource): Ditto.
(WebKit::RemoteRealtimeMediaSource::createRemoteMediaSource): Ditto.
(WebKit::toSourceType): Deleted.
(WebKit::m_manager): Deleted.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::createRemoteMediaSource): Ditto.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeVideoSource.cpp:
(WebKit::RemoteRealtimeVideoSource::create): Ditto.
(WebKit::RemoteRealtimeVideoSource::RemoteRealtimeVideoSource): Ditto.
(WebKit::RemoteRealtimeVideoSource::clone): Ditto.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeVideoSource.h:

* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.cpp:
(WebKit::UserMediaCaptureManager::AudioFactory::createAudioCaptureSource): Ditto.
(WebKit::UserMediaCaptureManager::VideoFactory::createVideoCaptureSource): Ditto.
(WebKit::UserMediaCaptureManager::DisplayFactory::createDisplayCaptureSource): Ditto.
* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.h:

* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl: Add setMockMediaDeviceIsEphemeral.
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setMockMediaDeviceIsEphemeral):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setMockMediaDeviceIsEphemeral):
* Tools/WebKitTestRunner/TestController.h:

* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/256161@main">https://commits.webkit.org/256161@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0491eb790218c415be2b0c22a70f8af1c090b768

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94920 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4068 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27823 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104526 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164790 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98915 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4153 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32249 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87223 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100447 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100589 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3000 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/81499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29988 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72875 "Found 1 new API test failure: /TestWebKit:WebKit.ParentFrame (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/86061 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38661 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18299 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/81285 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36493 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19580 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28000 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4247 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40418 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/81499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83957 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42394 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38812 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18977 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
```diff
@@ -70,15 +73,15 @@ RealtimeMediaSourceCenter::~RealtimeMediaSourceCenter() = default;
     RefPtr<RealtimeMediaSource> audioSource;
     if (audioDevice) {
-        auto source = audioCaptureFactory().createAudioCaptureSource(WTFMove(audioDevice), String { hashSalt }, &request.audioConstraints, request.pageIdentifier);
+        auto source = audioCaptureFactory().createAudioCaptureSource(WTFMove(audioDevice), MediaDeviceHashSalts { hashSalts }, &request.audioConstraints, request.pageIdentifier);
```
Why isn't this just a `WTFMove(hashSalts)`?

eric-carlson replied:
Because we may also need to create a video source with the same hash salts.

jernoble replied:
Aha! Makes sense!

jernoble commented on Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
```diff
@@ -205,7 +208,7 @@ void RealtimeMediaSourceCenter::getDisplayMediaDevices(const MediaStreamRequest&
             return;
 
-        auto sourceOrError = displayCaptureFactory().createDisplayCaptureSource(device, String { hashSalt }, &request.videoConstraints, request.pageIdentifier);
+        auto sourceOrError = displayCaptureFactory().createDisplayCaptureSource(device, MediaDeviceHashSalts { hashSalts }, &request.videoConstraints, request.pageIdentifier);
```
Ditto here and elsewhere.

eric-carlson replied:
These factory functions each return a vector of capture devices and the salts are needed to construct each one.

jernoble reviewed (approved):
r=me with nit:

eric-carlson committed:
Use an ephemeral device ID for the Continuity camera when it is not the system preferred
https://bugs.webkit.org/show_bug.cgi?id=247208
rdar://100335624

Reviewed by Jer Noble.

A "Continuity Camera", an iPhone used as a camera for a Macintosh, is listed as an
AVCaptureDevice whenever it is in close enough to the Macintosh to be used, but it is only
the "system preferred camera" when it is in an appropriate position and orientation to be
used as a camera for the device. This means that a script may remember the device ID when
a user has set up the phone to be used as a camera, and then pass it to getUserMedia later
when the phone is near the Macintosh but not usable as a camera - e.g. in the user's pocket.

To prevent this problem, use a device ID that is only valid for the lifetime of the
current frame when a Continuity Camera is available but is not the system preferred camera.

* LayoutTests/http/tests/media/media-stream/enumerate-devices-ephemeral-id-expected.txt: Added.
* LayoutTests/http/tests/media/media-stream/enumerate-devices-ephemeral-id.html: Added.
* LayoutTests/http/tests/media/media-stream/resources/enumerate-devices-ephemeral-id-iframe.html: Added.

* Source/WTF/wtf/PlatformHave.h: Define HAVE_CONTINUITY_CAMEARA

* Source/WebCore/Modules/mediastream/CanvasCaptureMediaStreamTrack.cpp:
(WebCore::CanvasCaptureMediaStreamTrack::Source::Source): Update for RealtimeMediaSource constructor
changes.
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::MediaDevices::exposeDevices): Take a MediaDeviceHashSalts instead of a String
because there are now two hash salts.
(WebCore::MediaDevices::enumerateDevices): Ditto.
* Source/WebCore/Modules/mediastream/MediaDevices.h:

* Source/WebCore/Modules/mediastream/UserMediaClient.h: Ditto.
* Source/WebCore/Modules/mediastream/UserMediaController.cpp:
* Source/WebCore/Modules/mediastream/UserMediaController.h:
(WebCore::UserMediaController::enumerateMediaDevices): Ditto.

* Source/WebCore/Modules/mediastream/UserMediaRequest.cpp:
(WebCore::UserMediaRequest::allow): Ditto.
* Source/WebCore/Modules/mediastream/UserMediaRequest.h:

* Source/WebCore/Modules/speech/SpeechRecognitionCaptureSource.cpp:
(WebCore::SpeechRecognitionCaptureSource::createRealtimeMediaSource): Update for
RealtimeMediaSource constructor changes.

* Source/WebCore/Modules/webaudio/MediaStreamAudioSource.cpp:
(WebCore::MediaStreamAudioSource::MediaStreamAudioSource): Ditto.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/platform/mediastream/CaptureDevice.h:
(WebCore::CaptureDevice::CaptureDevice): Add `isEphemeral` parameter.
(WebCore::CaptureDevice::setPersistentId):
(WebCore::CaptureDevice::setLabel):
(WebCore::CaptureDevice::isEphemeral const):
(WebCore::CaptureDevice::setIsEphemeral):
(WebCore::CaptureDevice::encode const):
(WebCore::CaptureDevice::decode):
(WebCore::CaptureDevice::isolatedCopy):

* Source/WebCore/platform/mediastream/RealtimeIncomingAudioSource.cpp:
(WebCore::RealtimeIncomingAudioSource::RealtimeIncomingAudioSource): Update for
RealtimeMediaSource constructor changes.

* Source/WebCore/platform/mediastream/RealtimeIncomingVideoSource.cpp:
(WebCore::RealtimeIncomingVideoSource::RealtimeIncomingVideoSource): Ditto.

* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::toSourceType):
(WebCore::RealtimeMediaSource::RealtimeMediaSource): Take a CaptureDevice instead of a
type, name, and device ID. Take a MediaDeviceHashSalts instead of a String for the salts.
(WebCore::RealtimeMediaSource::setPersistentId):
(WebCore::RealtimeMediaSource::initializePersistentId): Calculate persistent and ephemeral
hashed IDs.
(WebCore::RealtimeMediaSource::fitnessDistance): m_hashedID -> hashedID()
(WebCore::RealtimeMediaSource::hashedId const): Consider ephemeral state.
(WebCore::RealtimeMediaSource::deviceIDHashSalts const):
(WebCore::RealtimeMediaSource::setLogger):
(WebCore::RealtimeMediaSource::deviceIDHashSalt const): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:

* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::createMediaStream): Take MediaDeviceHashSalts with the
hash salts.
(WebCore::RealtimeMediaSourceCenter::getDisplayMediaDevices): Ditto.
(WebCore::RealtimeMediaSourceCenter::getUserMediaDevices): Ditto.
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraints): Ditto.
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraintsAfterEnumeration): Ditto.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.h:

* Source/WebCore/platform/mediastream/RealtimeMediaSourceFactory.h: Ditto.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h: Ditto.

* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::RealtimeVideoCaptureSource): Update for RealtimeMediaSource
constructor changes.
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:

* Source/WebCore/platform/mediastream/RealtimeVideoSource.cpp:
(WebCore::RealtimeVideoSource::RealtimeVideoSource): Ditto.

* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::create): Ditto.
(WebCore::DisplayCaptureSourceCocoa::DisplayCaptureSourceCocoa): Ditto.
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.h:

* Source/WebCore/platform/mediastream/ios/CoreAudioCaptureSourceIOS.h:
* Source/WebCore/platform/mediastream/ios/CoreAudioCaptureSourceIOS.mm:
(WebCore::CoreAudioCaptureSourceFactoryIOS::createAudioCaptureSource): Ditto.

* Source/WebCore/platform/mediastream/mac/AVCaptureDeviceManager.mm:
(WebCore::AVCaptureDeviceManager::updateCachedAVCaptureDevices):
(WebCore::toCaptureDevice): Make Continuity Camera ephemeral when it is not the system
preferred camera.
(WebCore::AVCaptureDeviceManager::retrieveCaptureDevices):

* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::create): Update for RealtimeMediaSource constructor changes.
(WebCore::AVVideoCaptureSource::AVVideoCaptureSource): Ditto.

* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSource::create): Ditto.
(WebCore::CoreAudioCaptureSource::createForTesting): Ditto.
(WebCore::CoreAudioCaptureSource::CoreAudioCaptureSource): Ditto.
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
(WebCore::CoreAudioCaptureSourceFactory::createAudioCaptureSource):

* Source/WebCore/platform/mediastream/mac/MockAudioSharedUnit.mm:
(WebCore::MockRealtimeAudioSource::create): Ditto.

* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.h:
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
(WebCore::MockRealtimeVideoSource::create): Ditto.
(WebCore::MockRealtimeVideoSourceMac::createForMockDisplayCapturer): Ditto.
(WebCore::MockRealtimeVideoSourceMac::MockRealtimeVideoSourceMac): Ditto.

* Source/WebCore/platform/mediastream/mac/RealtimeMediaSourceCenterMac.cpp:
* Source/WebCore/platform/mock/MockMediaDevice.h:
(WebCore::MockMediaDevice::captureDevice const): Allow to be ephemeral .
(WebCore::MockMediaDevice::encode const):
(WebCore::MockMediaDevice::decodeMockMediaDevice):
(WebCore::MockMediaDevice::decode):

* Source/WebCore/platform/mock/MockRealtimeAudioSource.cpp:
(WebCore::MockRealtimeAudioSource::create): Update for RealtimeMediaSource constructor changes.
(WebCore::MockRealtimeAudioSource::MockRealtimeAudioSource): Ditto.
* Source/WebCore/platform/mock/MockRealtimeAudioSource.h:

* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices): No devices are ephemeral by default.
(WebCore::MockDisplayCapturer::MockDisplayCapturer): Update for RealtimeMediaSource constructor
changes.
(WebCore::MockRealtimeMediaSourceCenter::setDeviceIsEphemeral):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.h:

* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::create): Ditto.
(WebCore::MockRealtimeVideoSource::MockRealtimeVideoSource): Ditto.
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::setMockMediaDeviceIsEphemeral): Added for testing.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/UIProcess/API/C/WKMockMediaDevice.cpp:
(WKAddMockMediaDevice):
(WKSetMockMediaDeviceIsEphemeral):
* Source/WebKit/UIProcess/API/C/WKMockMediaDevice.h:

* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::createMicrophoneSource):
(WebKit::UserMediaCaptureManagerProxy::createCameraSource):
(WebKit::UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::setMockMediaDeviceIsEphemeral):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/SpeechRecognitionRemoteRealtimeMediaSource.cpp:
(WebKit::SpeechRecognitionRemoteRealtimeMediaSource::SpeechRecognitionRemoteRealtimeMediaSource):
(): Deleted.
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest): Take MediaDeviceHashSalts
for hash salts.
(WebKit::UserMediaPermissionRequestManagerProxy::didCommitLoadForFrame): Clear ephemeral
salt for a frame.
(WebKit::UserMediaPermissionRequestManagerProxy::resetAccess): Ditto.
(WebKit::UserMediaPermissionRequestManagerProxy::ephemeralDeviceHashSaltForFrame): Calculate
or lookup an ephemeral salt for a frame.
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionRequest): Use
MediaDeviceHashSalts.
(WebKit::UserMediaPermissionRequestManagerProxy::platformValidateUserMediaRequestConstraints): Ditto.
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionValidRequest): Ditto.
(WebKit::UserMediaPermissionRequestManagerProxy::enumerateMediaDevicesForFrame): Ditto.
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:

* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
(WebKit::UserMediaPermissionRequestProxy::setDeviceIdentifierHashSalts): *HashSalt -> *HashSalts
(WebKit::UserMediaPermissionRequestProxy::deviceIdentifierHashSalts const): Ditto
(WebKit::UserMediaPermissionRequestProxy::setDeviceIdentifierHashSalt): Deleted.
(WebKit::UserMediaPermissionRequestProxy::deviceIdentifierHashSalt const): Deleted.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame): Let UserMediaPermissionRequestManager know the
frame has reloaded so it can clear the ephemeral salt ID.
(WebKit::WebPageProxy::enumerateMediaDevicesForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::setMockMediaDeviceIsEphemeral):
* Source/WebKit/UIProcess/WebProcessPool.h:

* Source/WebKit/WebProcess/MediaStream/UserMediaPermissionRequestManager.cpp:
(WebKit::UserMediaPermissionRequestManager::userMediaAccessWasGranted):
(WebKit::UserMediaPermissionRequestManager::enumerateMediaDevices):
* Source/WebKit/WebProcess/MediaStream/UserMediaPermissionRequestManager.h:

* Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.cpp:
(WebKit::WebUserMediaClient::enumerateMediaDevices):
* Source/WebKit/WebProcess/WebCoreSupport/WebUserMediaClient.h:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::userMediaAccessWasGranted):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setMockMediaDeviceIsEphemeral):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeAudioSource.cpp:
(WebKit::RemoteRealtimeAudioSource::create): Update for RealtimeMediaSource constructor change.
(WebKit::RemoteRealtimeAudioSource::RemoteRealtimeAudioSource): Ditto.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeAudioSource.h:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::RemoteRealtimeMediaSource): Ditto.
(WebKit::RemoteRealtimeMediaSource::createRemoteMediaSource): Ditto.
(WebKit::toSourceType): Deleted.
(WebKit::m_manager): Deleted.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::createRemoteMediaSource): Ditto.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeVideoSource.cpp:
(WebKit::RemoteRealtimeVideoSource::create): Ditto.
(WebKit::RemoteRealtimeVideoSource::RemoteRealtimeVideoSource): Ditto.
(WebKit::RemoteRealtimeVideoSource::clone): Ditto.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeVideoSource.h:

* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.cpp:
(WebKit::UserMediaCaptureManager::AudioFactory::createAudioCaptureSource): Ditto.
(WebKit::UserMediaCaptureManager::VideoFactory::createVideoCaptureSource): Ditto.
(WebKit::UserMediaCaptureManager::DisplayFactory::createDisplayCaptureSource): Ditto.
* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.h:

* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl: Add setMockMediaDeviceIsEphemeral.
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setMockMediaDeviceIsEphemeral):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setMockMediaDeviceIsEphemeral):
* Tools/WebKitTestRunner/TestController.h:

* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/256161@main
